### DISCUSSION TITLE
Storing data permanently as a file

### DISCUSSION THREAD

**Discussion by maibra88** [jspsych/jsPsych#2866] [General]
Hi all, I have an issue storing data as described here  https://www.jspsych.org/7.0/overview/data/, using a php script. I know that there are a bunch of posts n this topic out there, but 1. they all refer to older versions of jsPsych, and 2. none of the approaches helped me solve the puzzle. Below my code snippets, JS and PHP:

```
var jsPsych = initJsPsych({
    show_progress_bar: true,
    on_finish: function () {
      saveData(name + '.json', jsPsych.data.get().json());
    }
  })

  function saveData(name, data) {
    console.log(data)
    var xhr = new XMLHttpRequest();
    xhr.open('POST', '../../store.php'); // 'write_data.php' is the path to the php file described above.
    xhr.setRequestHeader('Content-Type', 'application/json');
    xhr.send(JSON.stringify({filename:name, filedata: data }));
  }
```

php:
```
<?php
// php for saving the data 
// set $debug to true in order to display error messages, you normally don't want this enabled on the server
$debug = true;
if ($debug) {
  error_reporting(E_ALL & ~E_NOTICE);
  ini_set('display_errors', 1);
  ini_set('display_startup_errors', 1);
}

// retreive content
$post_data = json_decode(file_get_contents('php://input'), true);
// write the file to disk
$name = "data/" . $post_data['filename'] . ".json";
$data = $post_data['filedata'];
file_put_contents($name, $data);

```

I host my experiment over the university server, and they have a set up a server to store the data. Hence, the data file into which the data should be saved is not in the git repo, but some external directory. I am not sure whether the problem lies with my script, or with the server where the data is supposed to be saved? I would really appreciate some insights. Thank you!

**Comment by ChristopheBossens** [comment]
Note sure what is happening, but I would try the following to get a better grip on the problem:

### 1. In your JavaScript saveData function, add the following after creating the xhr object:

```
xhr.onreadystatechange = function() {
        if (this.readyState == 4 ) {
            console.log(this.responseText);
        }
    };
```

This will print any info that the server produces when running the php script in the console window of your browser, and might contain useful information/error messages.

### Create a PhP script that simply saves some data and visit this script in the browsre

```
$data = "test";
file_put_contents("data_file.txt", $data);
```

That way you can check if it is related to how you are parsing the data sent through jsPsych, or if there is an issue on the server itself (perhaps some permissions need to be set before you can actually write data)?
